### DISCUSSION TITLE
[RFC] Structured data for non-built in commands

### DISCUSSION THREAD

**Discussion by KoviRobi** [xonsh/xonsh#5076] [Q&A]
Hi,

I am looking at coming up with a portable way to tell programs that they can output structured data, and telling the shell that the said data is structured according to the given format. For this I have a draft proposal write-up at https://github.com/KoviRobi/structured-shell-stdio but I would like some feedback.

This should avoid having to define aliases for all the commands that support e.g. JSON, for example
```
import json
json.loads($(ip --json addr))
```
could become just
```
ip addr
```
(assuming also some pretty-printing for structured data, such as [nushell](https://nushell.sh/) does, which also has `explore` which is a pager like `less` but supports structured data. Sorry I'm not very familiar with xonsh)

Thanks,
Robert

**Comment by anki-code** [comment]
In xonsh you can create [ExecAlias or callable alias](https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#aliases) with the logic you need or transform the command before execution using [on_transform_command](https://xon.sh/events.html#on-transform-command) event in interactive mode. I think xonsh is the best of the worlds for experiments.

**Reply by KoviRobi** [reply]
Thank you, yeah I agree such solutions work well, my goal was to reach out to see if anyone else had thought of this, and if so to try and come up with a single portable solution, and to act as a place to nucleate potential solutions and discussions. Also yeah I have to agree xonsh seems good for experiments, and a cool software :)
